File: primes/primes.rb
class Primes
  def self.prime?(number)
    return false if number <= 1
    return true if number == 2
    maximum = Math.sqrt(number).to_i + 1
    (2..maximum).each do |i|
      return false if number % i == 0
    end
    true
  end

  def self.list(size)
    answer = []
    return answer if size < 1
    generator = Primes::Generator.new
    size.times { answer << generator.next }
    answer
  end

  class Generator
    def initialize
      @last_prime = nil
      @next_candidate = nil
    end

    def next
      if @last_prime.nil?
        @next_candidate = 3
        @last_prime = 2
      else
        while !Primes.prime?(@next_candidate)
          @next_candidate += 2
        end
        answer = @next_candidate
        @next_candidate += 2
        answer
      end
    end
  end
end

File: primes/primes.py
"""This script gives basic prime number functions."""

import math
import itertools


def is_prime(number):
    """Checks if a number is prime"""
    if number <= 1:
        return False
    if number == 2:
        return True
    else:
        maximum = int(math.sqrt(number)) + 1
        for i in range(2, maximum):
            if number % i == 0:
                return False
        return True


def primes_gen():
    """Generator of prime numbers"""
    for number in itertools.count(2):
        if is_prime(number):
            yield number


def primes_list(size):
    """Generates a list of n prime numbers"""
    return list(itertools.islice(primes_gen(), size))


def prime_decomposition(number):
    """Returns a dictionary with the prime decomposition of n"""
    decomposition = {}
    number = int(number)
    if number < 2:
        return decomposition
    gen = primes_gen()
    break_condition = int(math.sqrt(number))
    while number > 1:
        current_prime = next(gen)
        if current_prime > break_condition:
            decomposition[number] = 1
            return decomposition
        while number % current_prime == 0 or number == current_prime:
            if current_prime in decomposition:
                decomposition[current_prime] += 1
            else:
                decomposition[current_prime] = 1
            number /= current_prime
    return decomposition


def get_from_decomposition(decomposition):
    """Returns a number from a prime decomposition"""
    result = 1
    for key in decomposition:
        result *= math.pow(key, decomposition[key])
    return result
